### ISSUE TITLE
Codegen issue with SSE2.Store interactions

### ISSUE THREAD

**Issue by redknightlois** [dotnet/coreclr#17435]
We found this issue on some of our code and managed to create a synthetic repro. The error happens because Sse2.Store does not invalidate the values already at the registers.

This method should return (3,6) but on release mode it returns (3,3). 
```csharp
static unsafe void Main(string[] args)
{
            (uint a, uint b) = XXHash32.Repro();
            Console.WriteLine($"{a}, {b}");
}

[MethodImpl(MethodImplOptions.NoInlining)]
public static (uint, uint) Repro()
{
            uint* a = stackalloc uint[4];
            a[0] = 1;
            a[1] = 1;
            a[2] = 1;
            a[3] = 1;

            // Here we force populate the registers
            var b = a[0];
            var h = a[3];
            var c = a[1];
            
            // We operate the values in xmm0
            Vector128<uint> v = Sse2.LoadVector128(a);
            v = Sse2.Add(v, v);
            // We send to the memory the modified values
            Sse2.Store(a, v);

            // We return both sums (from registers and from memory)
            return (b + h + c, a[0]+a[1]+a[3]);
}
```
On debug the behavior is correct because it is forced to get it from the memory location. We tested on 2.1.0-preview3-26404-01 

cc @fiigii @tannergooding 
Looping in @karelz because of ZBB

**Comment by fiigii** [comment]
Thank you for opening this issue. It looks like a general issue that JIT does not detect the "unsafe" memory write. Could you try to replace `Sse2.Store` by `Unsafe.Write` to see if it has this issue also?

cc @CarolEidt @AndyAyersMS

**Comment by mikedn** [comment]
> Thank you for opening this issue. It looks like a general issue that JIT does not detect the "unsafe" memory write

It's more likely that it's because the part about adding special handling in VN and other places for intrinsic load/stores was never done...

**Comment by redknightlois** [comment]
I tend to agree with @mikedn because ```Unsafe.Write``` works flawlessly.

**Comment by karelz** [comment]
FYI: @redknightlois ZBB is finished :) ... if you think this is important for 2.1, please work with area owners & experts - @CarolEidt @AndyAyersMS etc.

**Comment by briansull** [comment]
I have the following repro case that I am trying to compile:

```
using System;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using System.Runtime.Intrinsics.X86;
using System.Runtime.Intrinsics;

namespace Repro
{
    class Program
    {
        static unsafe void Main(string[] args)
        {
            (uint a, uint b) = Program.Repro();
            Console.WriteLine($"{a}, {b}");
        }

        [MethodImpl(MethodImplOptions.NoInlining)]
        public unsafe static (uint, uint) Repro()
        {
            uint* a = stackalloc uint[4];
            a[0] = 1;
            a[1] = 1;
            a[2] = 1;
            a[3] = 1;
            
            // Here we force populate the registers
            var b = a[0];
            var h = a[3];
            var c = a[1];
            
            // We operate the values in xmm0
            Vector128<uint> v = Sse2.LoadVector128(a);
            v = Sse2.Add(v, v);
            // We send to the memory the modified values
            Sse2.Store(a, v);
            
            // We return both sums (from registers and from memory)
            return (b + h + c, a[0]+a[1]+a[3]);
        }
    }
}

```

**Comment by briansull** [comment]
using:

```
> cd bin\Product\Windows_NT.x64.Checked
> "copy repro.cs here"
> csc /unsafe /o /r:System.Numerics.dll /r:System.Runtime.Intrinsics.Experimental.dll repro.cs
Microsoft (R) Visual C# Compiler version 2.2.0.61624
Copyright (C) Microsoft Corporation. All rights reserved.

repro.cs(32,13): error CS0246: The type or namespace name 'Vector128<>' could not be found (are you missing a using directive or an assembly reference?)
repro.cs(32,33): error CS0103: The name 'Sse2' does not exist in the current context
repro.cs(33,17): error CS0103: The name 'Sse2' does not exist in the current context
repro.cs(35,13): error CS0103: The name 'Sse2' does not exist in the current context

```

**Comment by briansull** [comment]
@redknightlois @fiigi @dotnet/jit-contrib
 Please advice on how to successfully build a stand alone repro for this Sse2 issue?

**Comment by AndyAyersMS** [comment]
This compiles for me with your program text -- note for this you need to have built the test overlay. You can probably do something similar with reference assemblies....
```
csc /unsafe /o /nostdlib /noconfig repro.cs
/r:d:\repos\coreclr\bin\Product\Windows_NT.x64.Checked\system.private.corelib.dll
/r:d:\repos\coreclr\bin\tests\Windows_NT.x64.Checked\tests\core_root\system.console.dll
/r:d:\repos\coreclr\bin\tests\Windows_NT.x64.Checked\tests\core_root\system.runtime.dll
```
and it repros the bug:
```
3, 3
```

**Comment by briansull** [comment]
THe SIMD store intrinsic treats the target as an SSA use rather than a def:

```
***** BB01, stmt 11
     (  3,  3) [000086] ------------              *  STMT      void  (IL   ???...0x05C)
N002 (  1,  1) [000084] ------------              |  /--*  LCL_VAR   simd16 V03 loc3         u:4 (last use) $285
N003 (  3,  3) [000085] ------------              \--*  HWIntrinsic void   uint Store $c3
N001 (  1,  1) [000083] ------------                 \--*  LCL_VAR   long   V00 loc0         u:3 $140
```

**Comment by mikedn** [comment]
@briansull That's correct, there's no actual target. There are 2 operands - the value and the store address - and they're both uses.

The problem is that nothing in the JIT knows that this node will actually store to memory. The net result is that a[0] & co. get CSEd because VN doesn't realize that the memory changed in between occurrences.

Such HW intrinsic node will need special handling, similar to what we have for `GT_LOCKADD`. Though a quick attempt at `GT_HWINTRINSIC` here and there yielded no improvement so I still don't know all the places in the JIT that need to be updated.

**Comment by briansull** [comment]
I believe that we have to expandthe Store operation into an Assignment like this:

```
               [000086] ------------              *  STMT      void  (IL   ???...  ???)
               [000084] ------------              |  /--*  LCL_VAR   simd16 V03 loc3         
               [000085] -A-XG-------              \--*  ASG       simd16 (copy)   
               [000084] *--XG-------                 /--*  IND       simd16   
               [000083] ------------                     \--*  LCL_VAR   long   V00 loc0         


```

**Comment by mikedn** [comment]
> I believe that we have to expandthe Store operation into an Assignment like this:

Ideally yes but unfortunately there are different kinds of store intrinsics and there are problems representing them using normal indirs. There were long discussions about intrinsic load/store representation (e.g. in #15991) but I don't know where it all landed. It looks like at some point a decision was taken to use GT_HWINTRINSIC for load/stores as well but I don't remember when or where that happened. 

Perhaps @fiigii or @CarolEidt can clear things up.

**Comment by briansull** [comment]
It would be pretty messy to add a new kind of Assignment node to the JIT.

Is is possible to change the signatures for these methods to use ref semantics?

` public static unsafe void Store(sbyte* address, Vector128<sbyte> source) => Store(address, source);`
to
` public static unsafe void Store(ref sbyte address, Vector128<sbyte> source) => Store(address, source);`

**Comment by CarolEidt** [comment]
@briansull - how would that help? It would seem that whatever IL you generate for those could be generated as easily for the pointer version. What am I missing?

**Comment by briansull** [comment]
I have worked up something that fixes the issue that we have here, but I am not confident that it is the correct approach.  In my fix any LclVar that is used as a SIMD operand to a load or a store is marked as 'lvIsSIMDMemortAddr' and then it is excluded from SSA.  This isn't exactly correct since these address are not actually changing but we are missing the Indirection node that should be associated with their uses, so this acts a subsitute for that.

**Comment by CarolEidt** [comment]
The issue of ref vs. pointer parameters was discussed in some detail in the API discussions, so I would propose that we not reopen that discussion until we've had some feedback from the preview.

> In my fix any LclVar that is used as a SIMD operand to a load or a store is marked as 'lvIsSIMDMemortAddr' and then it is excluded from SSA.

I assume you are just talking about the indirection operand, not the source operand in a store.

However, I don't think that will really fix the broader issue, because you could have a load or store whose address operand was not immediately visible as the address of a specific local.

Would it make sense for SSA to treat these loads/stores as reads/writes of any address-taken locals?

**Comment by tannergooding** [comment]
I believe we had decided to use `GT_HWIntrinsic` because there was a number of different states/other requirements that would not be representable by the current `GT_IND` nodes.

That is, we have the contract to emit a particular instruction, plus various modes such as:
* Scalar Load/Stores
* Vector Load/Stores
* Partial or Masked Load/Stores
* Temporal Load/Stores
* etc

IIRC, there was some concerns over the number of additional flags required to represent all the various states and that recognizing GT_HWIntrinsic nodes with the MemoryLoad/MemoryStore flag may be easier/more efficient overall.

**Comment by fiigii** [comment]
Sorry for the delay to reply. **I think the current best approach is to improve `GT_HWIntrinsic`'s load/store semantics even through that requires much effort.** We have discussed expanding Load/Store intrinsics to Ind/StoreInd in many places like https://github.com/dotnet/coreclr/pull/15991. However, `GT_IND/GT_STOREIND` nodes have no room to encode the modes of load/store intrinsics, and we will add more guys in the future (e.g., AVX-512). Moreover, I do not know if `GT_IND/GT_STOREIND` is able to express the semantics of some special load/store, such as
```csharp
// 128-bit load for Vector256
public static unsafe Vector256<sbyte> InsertVector128(Vector256<sbyte> value, sbyte* address, byte index);

// 128-bit store for Vector256
public static unsafe void ExtractVector128(byte* address, Vector256<byte> value, byte index);
```

I will look into this issue and try to provide a fix. Coulld you guys give me any hint about how `GT_STOREIND` or `Unsafe.Write` address this issue? @CarolEidt @AndyAyersMS @briansull @mikedn

**Comment by CarolEidt** [comment]
@fiigii - please do not work on a fix without having a discussion first about the requirements and the approach to take. This is a design issue, not simply a bug fix.

As I stated before, I would like to see a specification of the requirements *before* we even begin to talk about a solution. I still haven't seen that.

I still believe (without having seen evidence to the contrary, nor actually implementing this approach) that the correct approach is to use an addressing mode to capture the special addressing semantics of an intrinsic load or store. This would capture any special semantics for the memory operation itself. An additional intrinsic node may be required to convey any operation semantics, such as masking or merging involving registers.

This is not without front-end cost, as the front-end of the JIT is not used to seeing addressing modes - they are generally introduced by Lowering. However, the investment in making that work will pay off in addressing #5526.

It may be that a short-term solution for 2.1 can be found that will have lower cost and address the currently exposed load/store intrinsics.

**Comment by CarolEidt** [comment]
I had a discussion with @briansull and @tannergooding and I believe this is the best way forward:
- Make a new member of `genTreeOps` called something like `GT_HW_ADDR_MODE`. These would be instances of `GenTreeAddrMode`, which carries a base, an optional index, and a scale and offset, and is currently used only for `GT_LEA`.
- Any special characteristics of the addressing mode (non-temporal, aligned/unaligned, etc.) would be special addressing mode flags on that node, and thus would not clutter the already crowded `GT_IND` space.
- In the importer:
  - A store would be imported as `GT_ASG(GT_IND(GT_HW_ADDR_MODE(base,...)), value)`
  - A load would be imported as `GT_IND(GT_HW_ADDR_MODE(base,...))`
  - Any load or store with operational semantics (e.g. some kind of merge) would be implemented as a separate node above the `GT_IND`, and it would be up to the code generator, when generating that node, to handle the `GT_IND` child appropriately.

In the backend, the `GT_ASG(GT_IND))` would be transformed to a `GT_STOREIND` as per usual.

This will require some enhancements to the front-end to ensure that these nodes are handled properly, but I believe this is a more accurate capture of the semantics, and has less impact on non-hw-intrinsic code in the JIT.

There is indeed a gather instruction on x86 (`VGATHER*)`, that loads from multiple locations into a single vector register. That could perhaps be expressed as `GT_HWIntrinsic{gather}(GT_IND(GT_VECTOR_ADDR_MODE(address-vector)))`. I think we would probably want to use a different `gtOper` since the address isn't a normal type that we'd expect to see for an address, and there would probably be additional work required to enable this.

In the meantime, @briansull is working on a short-term fix for what we've got currently.

**Comment by briansull** [comment]
Fixed with #17464
